lib/logstash/outputs/kafka.rb

```diff
@@ -178,22 +195,80 @@ def register
         else
           record = org.apache.kafka.clients.producer.ProducerRecord.new(event.sprintf(@topic_id), event.sprintf(@message_key), data)
         end
-        @producer.send(record)
+        prepare(record)
       rescue LogStash::ShutdownSignal
         @logger.debug('Kafka producer got shutdown signal')
       rescue => e
         @logger.warn('kafka producer threw exception, restarting',
                      :exception => e)
       end
     end
-
   end # def register
 
-  def receive(event)
-    if event == LogStash::SHUTDOWN
-      return
+  def prepare(record)
+    # This output is threadsafe, so we need to keep a batch per thread.
+    @thread_batch_map[Thread.current].add(record)
+  end
+
+  def multi_receive(events)
+    t = Thread.current
+    if !@thread_batch_map.include?(t)
+      @thread_batch_map[t] = java.util.ArrayList.new(events.size)
+    end
+
+    events.each do |event|
+      break if event == LogStash::SHUTDOWN
+      @codec.encode(event)
+    end
+
+    batch = @thread_batch_map[t]
+    if batch.any?
+      retrying_send(batch)
+      batch.clear
     end
-    @codec.encode(event)
+  end
+
+  def retrying_send(batch)
+    remaining = @retries;
+
+    while batch.any?
+      if !remaining.nil?
+        if remaining < 0
+          # TODO(sissel): Offer to DLQ? Then again, if it's a transient fault,
+          # DLQing would make things worse (you dlq data that would be successful
+          # after the fault is repaired)
+          logger.info("Exhausted user-configured retry count when sending to Kafka. Dropping these events.",
+                      :max_retries => @retries, :drop_count => batch.count)
+          break
+        end
+
+        remaining -= 1
+      end
+
+      futures = batch.collect { |record| @producer.send(record) }
```

Comment:
@jordansissel one problem I see here is this:
The default `max.block.ms` (timeout on a `send` that can block if either the Kafka client's output buffer is full or fetching metadata is blocked) is `60s`. 
So for Kafka outages taking more than `max.block.ms` + (max metadata age) we'll start loosing data won't we (we will only catch these way upstream and just move on the next batch right now I think)?
I think we should catch these and retry the `send` calls with a back-off on `org.apache.kafka.common.errors.TimeoutException` right?

Reply:
Hmm.. Yeah, I didn't check what exceptions can be thrown here.

I'll add handling for the 3 listed here: https://kafka.apache.org/0100/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html

Reply:
Ok I added coverage for the 3 exceptions thrown by KafkaProducer.send() and added test coverage for it as well.